File: src/openrgb_flowers/service/runner_service.py
"""Execution service coordinating the main loop, frame pacing, and signals."""
from __future__ import annotations
import logging
import signal
import sys
import time
from typing import Optional

from openrgb_flowers.core.interfaces.i_effect_engine import IEffectEngine
from openrgb_flowers.core.interfaces.i_frame_transmitter import IFrameTransmitter
from openrgb_flowers.core.interfaces.i_layout_provider import ILayoutProvider
from openrgb_flowers.core.interfaces.i_visualizer import IVisualizer
from openrgb_flowers.core.models.effect_config import EffectConfig

logger = logging.getLogger("openrgb_flowers.service")


class RunnerService:
    """Coordinates high-performance main loop with precise frame pacing and clean shutdown."""

    def __init__(
        self,
        engine: IEffectEngine,
        transmitter: IFrameTransmitter,
        layout_provider: ILayoutProvider,
        config: EffectConfig,
        visualizer: Optional[IVisualizer] = None,
    ) -> None:
        self._engine = engine
        self._transmitter = transmitter
        self._layout = layout_provider
        self._config = config
        self._visualizer = visualizer
        self._running = False

    def stop(self) -> None:
        """Signals the main loop to stop."""
        self._running = False

    def run(self, max_frames: Optional[int] = None) -> None:
        """Runs the main effect rendering loop.

        Args:
            max_frames: Optional limit of frames to render (useful for tests/benchmarks).
        """
        self._running = True

        # Signal handlers for clean shutdown
        def _sig_handler(sig, frame):
            self.stop()

        try:
            signal.signal(signal.SIGINT, _sig_handler)
            signal.signal(signal.SIGTERM, _sig_handler)
        except (ValueError, AttributeError):
            # Signal handling may differ on some environments/threads
            pass

        # Connect transmitter
        self._transmitter.connect()

        target_fps = self._config.fps
        frame_time = 1.0 / target_fps
        prev_time = time.perf_counter()
        frames_rendered = 0

        logger.info(f"Starting Flowers Blooming effect loop @ {target_fps:.1f} FPS...")

        try:
            while self._running:
                loop_start = time.perf_counter()
                dt = loop_start - prev_time
                prev_time = loop_start

                # Clamp dt to avoid huge leap on lag spike
                dt_clamped = min(0.1, max(0.001, dt))

                # Advance simulation and render frame
                frame = self._engine.tick(dt_clamped)

                # Transmit frame to hardware
                self._transmitter.send_frame(frame)

                # Optional live visualizer
                if self._visualizer:
                    active_blooms = self._engine.get_active_flower_count()
                    self._visualizer.render(frame, self._layout, active_blooms, target_fps)

                frames_rendered += 1
                if max_frames and frames_rendered >= max_frames:
                    break

                # Precise frame throttling
                elapsed = time.perf_counter() - loop_start
                sleep_time = frame_time - elapsed
                if sleep_time > 0:
                    time.sleep(sleep_time)

        finally:
            logger.info("Shutting down Flowers Blooming effect...")
            if self._visualizer:
                self._visualizer.close()
            self._transmitter.disconnect()

File: src/service/runner_service.cpp
#include "openrgb_flowers/service/runner_service.hpp"
#include "openrgb_flowers/effects/effect_engine_factory.hpp"

#include <algorithm>
#include <chrono>
#include <csignal>
#include <iostream>
#include <thread>

namespace openrgb_flowers::service {

namespace {

static std::atomic<RunnerService*> s_active_instance{nullptr};

void signal_handler(int sig) {
    (void)sig;
    if (auto* inst = s_active_instance.load()) {
        inst->stop();
    }
}

} // anonymous namespace

RunnerService::RunnerService(
    std::shared_ptr<core::interfaces::IEffectEngine> engine,
    std::shared_ptr<core::interfaces::IFrameTransmitter> transmitter,
    std::shared_ptr<core::interfaces::ILayoutProvider> layout,
    core::models::EffectConfig config
) : engine_(std::move(engine)),
    transmitter_(std::move(transmitter)),
    layout_(std::move(layout)),
    config_(std::move(config)) {}

void RunnerService::stop() {
    running_ = false;
}

bool RunnerService::is_running() const {
    return running_;
}

void RunnerService::set_engine(std::shared_ptr<core::interfaces::IEffectEngine> engine) {
    std::lock_guard<std::mutex> lock(mutex_);
    engine_ = std::move(engine);
}

void RunnerService::set_palette(const std::string& palette_name) {
    std::lock_guard<std::mutex> lock(mutex_);
    config_.palette_name = palette_name;
    if (engine_) {
        engine_->update_config(config_);
    }
}

void RunnerService::set_effect(const std::string& effect_name) {
    std::lock_guard<std::mutex> lock(mutex_);
    config_.effect_type = effect_name;
    engine_ = effects::EffectEngineFactory::create_engine(effect_name, config_, layout_);
}

void RunnerService::update_config(const core::models::EffectConfig& config) {
    std::lock_guard<std::mutex> lock(mutex_);
    bool effect_changed = (config.effect_type != config_.effect_type);
    config_ = config;
    if (effect_changed) {
        engine_ = effects::EffectEngineFactory::create_engine(config_.effect_type, config_, layout_);
    } else if (engine_) {
        engine_->update_config(config_);
    }
}

void RunnerService::run(std::optional<uint64_t> max_frames) {
    running_ = true;
    s_active_instance.store(this);
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    if (!transmitter_->is_connected()) {
        transmitter_->connect();
    }

    auto prev_time = std::chrono::steady_clock::now();
    uint64_t frames_rendered = 0;

    while (running_) {
        auto loop_start = std::chrono::steady_clock::now();
        std::chrono::duration<float> dt_duration = loop_start - prev_time;
        prev_time = loop_start;

        float dt = dt_duration.count();
        float dt_clamped = std::clamp(dt, 0.001f, 0.1f);

        core::models::RenderFrame frame;
        float current_fps = 30.0f;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (engine_) {
                frame = engine_->tick(dt_clamped);
            }
            current_fps = config_.fps;
        }

        if (frame.led_count > 0) {
            transmitter_->send_frame(frame);
        }

        ++frames_rendered;
        if (max_frames.has_value() && frames_rendered >= max_frames.value()) {
            break;
        }

        auto frame_duration = std::chrono::duration<double>(1.0 / std::max(1.0f, current_fps));
        auto elapsed = std::chrono::steady_clock::now() - loop_start;
        auto sleep_time = frame_duration - elapsed;
        if (sleep_time > std::chrono::duration<double>::zero()) {
            std::this_thread::sleep_for(sleep_time);
        }
    }

    transmitter_->disconnect();
    s_active_instance.store(nullptr);
}

} // namespace openrgb_flowers::service
